### PR TITLE
Fix passing object ref to generateTestHook

### PR DESCRIPTION
**Pull request by MrLoh**
PR for #197

### CODE REVIEW

**Comment by MrLoh** [comment]
Not sure how to add a test for this, would be happy for some pointer

**Comment by AbigailMcP** [comment]
Hi @MrLoh ! Thanks for opening this PR and your work diagnosing the issue.

I've had some time to take a look over this today and in doing so have added a test to ensure that Cavy respects the functionality of the `useRef` hook.

You can run the tests yourself using the Tester app by following the instructions in it's [own README](https://github.com/pixielabs/cavy/blob/master/test/CavyTester/README.md) - it's just like running any other React Native app, but linking your local version of Cavy.

At the moment the `useRef` test fails with the error `Cannot read property 'focus' of undefined`.
This is because in my test, I use `useRef` as such:

```javascript
const textInput = useRef();
```
so `ref.current` is initialized to undefined. Therefore, in your code, `ref.current` is never set to the component:

```javascript
if (ref && ref.current) {
  ref.current = component;
}
```

**Comment by MrLoh** [comment]
@AbigailMcP thanks a lot for having a look I'll check the tests, a little confused why they are failing, what you are doing is exactly what we are doing in our codebase and we are depending on the forked package now. I'LLC habe a Look sometime this week.

**Comment by AbigailMcP** [comment]
> @AbigailMcP thanks a lot for having a look I'll check the tests, a little confused why they are failing, what you are doing is exactly what we are doing in our codebase and we are depending on the forked package now. I'LLC habe a Look sometime this week.

oh interesting! let me know what you find :)

File: test/CavyTester/src/scenarios/callbackRef.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import React, { useRef } from 'react';
```
Comment (AbigailMcP) [inline test/CavyTester/src/scenarios/callbackRef.js:1]: Please remove unused `useRef`.

File: test/CavyTester/src/scenarios/callbackRef.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import { Button, TextInput } from 'react-native';
+import { useCavy, hook } from 'cavy';
```
Comment (AbigailMcP) [inline test/CavyTester/src/scenarios/callbackRef.js:3]: As above! `useCavy` unused.

File: test/CavyTester/src/scenarios/createRef.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
+import React, { useRef } from 'react';
```
Comment (AbigailMcP) [inline test/CavyTester/src/scenarios/createRef.js:1]: As above, `useRef` unused.

File: test/CavyTester/src/scenarios/createRef.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import { Button, TextInput } from 'react-native';
+import { useCavy, hook } from 'cavy';
+
+export const key = 'CreateRef';
+const inputId = `${key}.Input`;
+const buttonId = `${key}.Button`;
+
+class CreateRefTest extends React.Component {
+  textInput = React.createRef();
```
Comment (AbigailMcP) [inline test/CavyTester/src/scenarios/createRef.js:10]: For style consistency, please could you import `createRef` in the same way as we do `useRef`?

Comment (AbigailMcP) [reply test/CavyTester/src/scenarios/createRef.js:10]: I'd also expect to see this `createRef()` in a constructor for this class component.

**Review by AbigailMcP — CHANGES_REQUESTED** [review]
A couple of style / tidy up comments left below :)

Also - great work extending our test suite! Thanks 🙏 

One question, that @jalada can help me with... how do we feel about string refs? The React documentation still describes them as legacy API and that they will be deprecated in an upcoming release, but they're still potentially in use. Do we want to continue supporting them?

If so, @MrLoh - would this change now break for people still using string refs?

**Comment by MrLoh** [comment]
@AbigailMcP maybe string refs is what the original implementation was build for. We can easily add another case to return the ref if it is a string. 

Tanks for the review. Will fix that stuff. You should really add a prettier and eslint config to the repo, to make sure these things are caught. I assume you have something like that setup internally. But since the repo doesn't have it, as an external contributor I cannot leverage it.

File: test/CavyTester/src/scenarios/createRef.js (hunk 4 above)
Comment (MrLoh) [reply test/CavyTester/src/scenarios/createRef.js:10]: Well this uses ESnext, no need for a constructor here. But I will refactor and also make sure binding is done manually then. I think it is bad practice to not use modern ESnext functionality fron binding, but that's just my opinion.

Comment (MrLoh) [reply test/CavyTester/src/scenarios/createRef.js:10]: I will also import Component separately from React then, otherwise this is still a mix and match

**Comment by MrLoh** [comment]
@AbigailMcP actually string refs are not supported in the current version and I don't see a way to support them, since for a string ref a string needs to be passed to the `ref` prop, but `generateTestHook` cannot be a string, since it is a function. The current untested functionality of returning a string from that function does nothing to support string refs.

**Comment by MrLoh** [comment]
@AbigailMcP I can actually add support for string refs in a rather hacky way. 

```js
  function generateTestHook(identifier, ref) {
    // support for legacy string refs
    if(typeof ref == 'string') {
      console.warn('string refs are deprecated and may not be supported in future versions of cavy. If you really need them, please use `generateTestHook.bind(this)(identifier, stringRef)`')
      setTimeout(() => {
        if(this.refs[ref]) {
          testHookStore.add(identifier, this.refs[ref]);
        } else {
          testHookStore.remove(identifier);
        }
      }, 1);
      return ref
    }

    ... current implementation
}
```

then you can use string refs by passing the following ref prop `ref={generateTestHook.bind(this)(inputId, 'textInput')}` but since string refs are not currently supported, I would say that is a separate issue and you propbably do not want to add support for a deprecated feature in a hacky way.

**Review by AbigailMcP — APPROVED** [review]

**Comment by AbigailMcP** [comment]
Thanks, agreed support for string ref is a separate issue. Will get this merged.

**Pull request merged**